src/main/java/io/cdap/plugin/jms/common/JMSConnection.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright © 2016-2020 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.jms.common;
+
+import com.google.common.base.Strings;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import java.util.Properties;
+import javax.jms.Connection;
+import javax.jms.ConnectionFactory;
+import javax.jms.Destination;
+import javax.jms.JMSException;
+import javax.jms.MessageConsumer;
+import javax.jms.MessageListener;
+import javax.jms.MessageProducer;
+import javax.jms.Queue;
+import javax.jms.Session;
+import javax.jms.Topic;
+import javax.naming.Context;
+import javax.naming.InitialContext;
+import javax.naming.NamingException;
+
+/**
+ * A facade class that encapsulates the necessary functionality to: get the initial context, resolve the connection
+ * factory, establish connection to JMS, create a session, resolve the destination by a queue or topic name, create
+ * producer, create consumer, set message listener to a consumer, and start connection. This class handles exceptions
+ * for all the functionalities provided.
+ */
+public class JMSConnection {
+
+    private static final Logger LOG = LoggerFactory.getLogger(JMSConnection.class);
+    private final JMSConfig config;
+
+    public JMSConnection(JMSConfig config) {
+        this.config = config;
+    }
+
+    public Context getContext() {
+        Properties properties = new Properties();
+        properties.put(Context.INITIAL_CONTEXT_FACTORY, config.getJndiContextFactory());
+        properties.put(Context.PROVIDER_URL, config.getProviderUrl());
+
+        if (!(Strings.isNullOrEmpty(config.getJndiUsername()) && Strings.isNullOrEmpty(config.getJndiPassword()))) {
+            properties.put(Context.SECURITY_PRINCIPAL, config.getJndiUsername());
+            properties.put(Context.SECURITY_CREDENTIALS, config.getJndiPassword());
+        }
+
+        try {
+            return new InitialContext(properties);
+        } catch (NamingException e) {
+            LOG.error("Initial context couldn't get created.", e);
+            throw new RuntimeException(e.getMessage());
```

Comment:
The cause should be passed together with a message that provide more context. E.g.

```
throw new RuntimeException("Failed to create initial context for provider URL " + config.getProviderUrl() + " with principal " + config.getJndiUsername(), e);
```

Reply:
Done